File: java/otto-crypt-java/src/main/java/com/ivansostarko/ottocrypt/Otto.java
package com.ivansostarko.ottocrypt;

import javax.crypto.Cipher;
import javax.crypto.KeyAgreement;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.*;
import java.security.spec.NamedParameterSpec;
import java.util.Arrays;

import static com.ivansostarko.ottocrypt.Utils.*;

public final class Otto {
    public static final int DEFAULT_CHUNK_SIZE = 1 << 20; // 1 MiB

    private Otto() {}

    // === Public API (in-memory message) ===

    public static OttoResult encryptString(byte[] plaintext, byte[] rawKey32) throws Exception {
        check(rawKey32 != null && rawKey32.length == 32, "rawKey32 must be 32 bytes");
        boolean chunked = false;
        byte[] fileSalt = random16();
        byte[] header = buildHeader(fileSalt, chunked);

        DerivedKeys dk = deriveKeys(rawKey32, fileSalt);
        byte[] nonce = deriveChunkNonce(dk.nonceKey, 0);
        byte[] cipher = encryptAesGcm(dk.encKey, nonce, header, plaintext);
        return new OttoResult(header, cipher);
    }

    public static byte[] decryptString(byte[] cipherAndTag, byte[] header, byte[] rawKey32) throws Exception {
        check(rawKey32 != null && rawKey32.length == 32, "rawKey32 must be 32 bytes");
        ParsedHeader ph = parseHeader(header);
        DerivedKeys dk = deriveKeys(rawKey32, ph.fileSalt);
        byte[] nonce = deriveChunkNonce(dk.nonceKey, 0);
        return decryptAesGcm(dk.encKey, nonce, header, cipherAndTag);
    }

    // === Files / streaming ===

    public static void encryptFile(Path input, Path output, byte[] rawKey32, int chunkBytes) throws Exception {
        check(rawKey32 != null && rawKey32.length == 32, "rawKey32 must be 32 bytes");
        byte[] fileSalt = random16();
        byte[] header = buildHeader(fileSalt, true);
        DerivedKeys dk = deriveKeys(rawKey32, fileSalt);

        try (InputStream in = Files.newInputStream(input);
             OutputStream out = Files.newOutputStream(output)) {
            out.write(header);
            byte[] buf = new byte[chunkBytes];
            long counter = 0;
            for (;;) {
                int n = in.read(buf);
                if (n < 0) break;
                if (n == 0) continue;
                byte[] pt = Arrays.copyOf(buf, n);
                byte[] nonce = deriveChunkNonce(dk.nonceKey, counter++);
                byte[] ctTag = encryptAesGcm(dk.encKey, nonce, header, pt);
                int ctLen = ctTag.length - TAG_LEN;
                out.write(u32be(ctLen));
                out.write(ctTag, 0, ctLen);
                out.write(ctTag, ctLen, TAG_LEN);
            }
            out.flush();
        }
    }

    public static void decryptFile(Path input, Path output, byte[] rawKey32) throws Exception {
        check(rawKey32 != null && rawKey32.length == 32, "rawKey32 must be 32 bytes");

        try (InputStream in = Files.newInputStream(input);
             OutputStream out = Files.newOutputStream(output)) {
            byte[] fixed = in.readNBytes(FIXED_HDR_LEN);
            check(fixed.length == FIXED_HDR_LEN, "bad header");
            check(Arrays.equals(Arrays.copyOfRange(fixed, 0, 5), MAGIC), "bad magic");
            check(fixed[5] == ALGO_ID && fixed[6] == KDF_RAW, "algo/kdf mismatch");
            int varLen = beU16(fixed, 9);
            byte[] var = in.readNBytes(varLen);
            check(var.length == varLen, "truncated header");
            byte[] header = ByteBuffer.allocate(FIXED_HDR_LEN + varLen).put(fixed).put(var).array();

            ParsedHeader ph = parseHeader(header);
            DerivedKeys dk = deriveKeys(rawKey32, ph.fileSalt);

            long counter = 0;
            for (;;) {
                byte[] lenb = in.readNBytes(4);
                if (lenb.length == 0) break;
                check(lenb.length == 4, "truncated chunk length");
                int clen = beU32(lenb, 0);
                if (clen == 0) break;
                byte[] ct = in.readNBytes(clen);
                check(ct.length == clen, "truncated ciphertext");
                byte[] tag = in.readNBytes(TAG_LEN);
                check(tag.length == TAG_LEN, "missing tag");

                byte[] nonce = deriveChunkNonce(dk.nonceKey, counter++);
                byte[] cat = ByteBuffer.allocate(ct.length + TAG_LEN).put(ct).put(tag).array();
                byte[] pt = decryptAesGcm(dk.encKey, nonce, header, cat);
                out.write(pt);
            }
            out.flush();
        }
    }

    // === X25519 helpers ===

    public static KeyPair x25519Generate() throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("X25519");
        kpg.initialize(new NamedParameterSpec("X25519"));
        return kpg.generateKeyPair();
    }
    public static byte[] x25519Shared(PrivateKey mySecret, PublicKey theirPublic) throws Exception {
        javax.crypto.KeyAgreement ka = javax.crypto.KeyAgreement.getInstance("X25519");
        ka.init(mySecret);
        ka.doPhase(theirPublic, true);
        return ka.generateSecret();
    }
    public static byte[] hkdfSession(byte[] shared, byte[] salt) throws Exception {
        return HKDF.derive(shared, salt, "OTTO-P2P-SESSION".getBytes(), 32);
    }

    // === internals ===

    private static byte[] buildHeader(byte[] fileSalt16, boolean chunked) {
        check(fileSalt16.length == FILE_SALT_LEN, "file salt must be 16 bytes");
        ByteBuffer buf = ByteBuffer.allocate(FIXED_HDR_LEN + FILE_SALT_LEN);
        buf.put(MAGIC);
        buf.put(ALGO_ID);
        buf.put(KDF_RAW);
        buf.put(chunked ? FLAG_CHUNKED : (byte)0x00);
        buf.put((byte)0x00);
        buf.put(u16be(FILE_SALT_LEN));
        buf.put(fileSalt16);
        return buf.array();
    }

    private static ParsedHeader parseHeader(byte[] header) {
        check(header != null && header.length >= FIXED_HDR_LEN, "header too short");
        check(Arrays.equals(Arrays.copyOfRange(header, 0, 5), MAGIC), "bad magic");
        check(header[5] == ALGO_ID, "algo mismatch");
        check(header[6] == KDF_RAW, "kdf mismatch");
        int varLen = beU16(header, 9);
        check(header.length == FIXED_HDR_LEN + varLen, "header length mismatch");
        check(varLen >= FILE_SALT_LEN, "missing file salt");
        byte[] fileSalt = Arrays.copyOfRange(header, FIXED_HDR_LEN, FIXED_HDR_LEN + FILE_SALT_LEN);
        boolean chunked = (header[7] & FLAG_CHUNKED) != 0;
        return new ParsedHeader(fileSalt, chunked);
    }

    private static final class ParsedHeader {
        final byte[] fileSalt;
        final boolean chunked;
        ParsedHeader(byte[] salt, boolean chunked) { this.fileSalt = salt; this.chunked = chunked; }
    }

    private static final class DerivedKeys {
        final byte[] encKey;   // 32
        final byte[] nonceKey; // 32
        DerivedKeys(byte[] e, byte[] n) { encKey = e; nonceKey = n; }
    }

    private static DerivedKeys deriveKeys(byte[] rawKey32, byte[] fileSalt) throws Exception {
        byte[] encKey = HKDF.derive(rawKey32, fileSalt, "OTTO-ENC-KEY".getBytes(), 32);
        byte[] nonceKey = HKDF.derive(rawKey32, fileSalt, "OTTO-NONCE-KEY".getBytes(), 32);
        return new DerivedKeys(encKey, nonceKey);
    }

    private static byte[] deriveChunkNonce(byte[] nonceKey32, long counter) throws Exception {
        byte[] info = ByteBuffer.allocate("OTTO-CHUNK-NONCE".length() + 8)
                .put("OTTO-CHUNK-NONCE".getBytes())
                .put(HKDF.be64(counter))
                .array();
        return HKDF.expandNonce(nonceKey32, info, NONCE_LEN);
    }

    private static byte[] encryptAesGcm(byte[] encKey32, byte[] nonce12, byte[] aad, byte[] pt) throws Exception {
        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKeySpec key = new SecretKeySpec(encKey32, "AES");
        GCMParameterSpec gcm = new GCMParameterSpec(8 * TAG_LEN, nonce12);
        c.init(Cipher.ENCRYPT_MODE, key, gcm);
        if (aad != null && aad.length > 0) c.updateAAD(aad);
        return c.doFinal(pt);
    }

    private static byte[] decryptAesGcm(byte[] encKey32, byte[] nonce12, byte[] aad, byte[] ctAndTag) throws Exception {
        Cipher c = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKeySpec key = new SecretKeySpec(encKey32, "AES");
        GCMParameterSpec gcm = new GCMParameterSpec(8 * TAG_LEN, nonce12);
        c.init(Cipher.DECRYPT_MODE, key, gcm);
        if (aad != null && aad.length > 0) c.updateAAD(aad);
        return c.doFinal(ctAndTag);
    }
}

File: android/ottocrypt-android/src/main/java/com/ivansostarko/ottocrypt/android/Otto.kt
package com.ivansostarko.ottocrypt.android

import android.os.Build
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.NamedParameterSpec
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import java.util.Arrays

import com.ivansostarko.ottocrypt.android.Utils.*

object Otto {
    const val DEFAULT_CHUNK_SIZE: Int = 1 shl 20

    @JvmStatic
    fun encryptString(plaintext: ByteArray, rawKey32: ByteArray): OttoResult {
        require(rawKey32.size == 32) { "rawKey32 must be 32 bytes" }
        val fileSalt = random16()
        val header = buildHeader(fileSalt, chunked = false)
        val (encKey, nonceKey) = deriveKeys(rawKey32, fileSalt)
        val nonce = deriveChunkNonce(nonceKey, 0)
        val cipher = encryptAesGcm(encKey, nonce, header, plaintext)
        return OttoResult(header, cipher)
    }

    @JvmStatic
    fun decryptString(cipherAndTag: ByteArray, header: ByteArray, rawKey32: ByteArray): ByteArray {
        require(rawKey32.size == 32) { "rawKey32 must be 32 bytes" }
        val ph = parseHeader(header)
        val (encKey, nonceKey) = deriveKeys(rawKey32, ph.fileSalt)
        val nonce = deriveChunkNonce(nonceKey, 0)
        return decryptAesGcm(encKey, nonce, header, cipherAndTag)
    }

    @JvmStatic
    fun encryptFile(input: File, output: File, rawKey32: ByteArray, chunkBytes: Int = DEFAULT_CHUNK_SIZE) {
        require(rawKey32.size == 32) { "rawKey32 must be 32 bytes" }
        val fileSalt = random16()
        val header = buildHeader(fileSalt, chunked = true)
        val (encKey, nonceKey) = deriveKeys(rawKey32, fileSalt)

        FileInputStream(input).use { fin ->
            FileOutputStream(output).use { fout ->
                fout.write(header)
                val buf = ByteArray(chunkBytes)
                var counter = 0L
                while (true) {
                    val n = fin.read(buf)
                    if (n < 0) break
                    if (n == 0) continue
                    val pt = Arrays.copyOf(buf, n)
                    val nonce = deriveChunkNonce(nonceKey, counter++)
                    val ctTag = encryptAesGcm(encKey, nonce, header, pt)
                    val ctLen = ctTag.size - TAG_LEN
                    fout.write(u32be(ctLen))
                    fout.write(ctTag, 0, ctLen)
                    fout.write(ctTag, ctLen, TAG_LEN)
                }
                fout.flush()
            }
        }
    }

    @JvmStatic
    fun decryptFile(input: File, output: File, rawKey32: ByteArray) {
        require(rawKey32.size == 32) { "rawKey32 must be 32 bytes" }
        FileInputStream(input).use { fin ->
            FileOutputStream(output).use { fout ->
                val fixed = fin.readNBytes(FIXED_HDR_LEN)
                require(fixed.size == FIXED_HDR_LEN) { "bad header" }
                require(fixed.copyOfRange(0,5).contentEquals(MAGIC)) { "bad magic" }
                require(fixed[5] == ALGO_ID && fixed[6] == KDF_RAW) { "algo/kdf mismatch" }
                val varLen = beU16(fixed, 9)
                val varp = fin.readNBytes(varLen)
                require(varp.size == varLen) { "truncated header" }
                val header = ByteBuffer.allocate(FIXED_HDR_LEN + varLen).put(fixed).put(varp).array()

                val ph = parseHeader(header)
                val (encKey, nonceKey) = deriveKeys(rawKey32, ph.fileSalt)

                var counter = 0L
                while (true) {
                    val lenb = fin.readNBytes(4)
                    if (lenb.isEmpty()) break
                    require(lenb.size == 4) { "truncated chunk length" }
                    val clen = beU32(lenb, 0)
                    if (clen == 0) break
                    val ct = fin.readNBytes(clen)
                    require(ct.size == clen) { "truncated ciphertext" }
                    val tag = fin.readNBytes(TAG_LEN)
                    require(tag.size == TAG_LEN) { "missing tag" }

                    val nonce = deriveChunkNonce(nonceKey, counter++)
                    val cat = ByteBuffer.allocate(ct.size + TAG_LEN).put(ct).put(tag).array()
                    val pt = decryptAesGcm(encKey, nonce, header, cat)
                    fout.write(pt)
                }
                fout.flush()
            }
        }
    }

    // X25519 (API 28+)
    @JvmStatic
    fun x25519Generate(): KeyPair? {
        if (Build.VERSION.SDK_INT < 28) return null
        val kpg = KeyPairGenerator.getInstance("X25519")
        kpg.initialize(NamedParameterSpec("X25519"))
        return kpg.generateKeyPair()
    }
    @JvmStatic
    fun x25519Shared(mySecret: PrivateKey, theirPublic: PublicKey): ByteArray {
        val ka = KeyAgreement.getInstance("X25519")
        ka.init(mySecret)
        ka.doPhase(theirPublic, true)
        return ka.generateSecret()
    }
    @JvmStatic
    fun hkdfSession(shared: ByteArray, salt: ByteArray): ByteArray =
        HKDF.derive(shared, salt, "OTTO-P2P-SESSION".toByteArray(), 32)

    private data class ParsedHeader(val fileSalt: ByteArray, val chunked: Boolean)

    private fun buildHeader(fileSalt16: ByteArray, chunked: Boolean): ByteArray {
        require(fileSalt16.size == FILE_SALT_LEN) { "file salt must be 16 bytes" }
        return ByteBuffer.allocate(FIXED_HDR_LEN + FILE_SALT_LEN).apply {
            put(MAGIC); put(ALGO_ID); put(KDF_RAW)
            put(if (chunked) FLAG_CHUNKED else 0x00); put(0x00)
            put(u16be(FILE_SALT_LEN)); put(fileSalt16)
        }.array()
    }
    private fun parseHeader(header: ByteArray): ParsedHeader {
        require(header.size >= FIXED_HDR_LEN) { "header too short" }
        require(header.copyOfRange(0,5).contentEquals(MAGIC)) { "bad magic" }
        require(header[5] == ALGO_ID) { "algo mismatch" }
        require(header[6] == KDF_RAW) { "kdf mismatch" }
        val varLen = beU16(header, 9)
        require(header.size == FIXED_HDR_LEN + varLen) { "header length mismatch" }
        require(varLen >= FILE_SALT_LEN) { "missing file salt" }
        val fileSalt = header.copyOfRange(FIXED_HDR_LEN, FIXED_HDR_LEN + FILE_SALT_LEN)
        val chunked = (header[7].toInt() and FLAG_CHUNKED.toInt()) != 0
        return ParsedHeader(fileSalt, chunked)
    }
    private data class DerivedKeys(val encKey: ByteArray, val nonceKey: ByteArray)
    private fun deriveKeys(rawKey32: ByteArray, fileSalt: ByteArray): DerivedKeys {
        val encKey = HKDF.derive(rawKey32, fileSalt, "OTTO-ENC-KEY".toByteArray(), 32)
        val nonceKey = HKDF.derive(rawKey32, fileSalt, "OTTO-NONCE-KEY".toByteArray(), 32)
        return DerivedKeys(encKey, nonceKey)
    }
    private fun deriveChunkNonce(nonceKey32: ByteArray, counter: Long): ByteArray {
        val label = "OTTO-CHUNK-NONCE".toByteArray()
        val info = ByteBuffer.allocate(label.size + 8).put(label).put(HKDF.be64(counter)).array()
        return HKDF.expandNonce(nonceKey32, info, NONCE_LEN)
    }
    private fun encryptAesGcm(encKey32: ByteArray, nonce12: ByteArray, aad: ByteArray, pt: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        val key = SecretKeySpec(encKey32, "AES")
        val gcm = GCMParameterSpec(8 * TAG_LEN, nonce12)
        c.init(Cipher.ENCRYPT_MODE, key, gcm)
        if (aad.isNotEmpty()) c.updateAAD(aad)
        return c.doFinal(pt)
    }
    private fun decryptAesGcm(encKey32: ByteArray, nonce12: ByteArray, aad: ByteArray, cat: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        val key = SecretKeySpec(encKey32, "AES")
        val gcm = GCMParameterSpec(8 * TAG_LEN, nonce12)
        c.init(Cipher.DECRYPT_MODE, key, gcm)
        if (aad.isNotEmpty()) c.updateAAD(aad)
        return c.doFinal(cat)
    }
}
